### PR TITLE
[AutoDiff] TF-123: fix reabstraction to opaque

### PR DESCRIPTION
**Pull request by marcrasi**
This fixes TF-123 by adding a new abstraction pattern that properly abstracts the derivative functions in a `differentiable_function` bundle that is being abstracted to `AP::Opaque`.

Even though this technically fixes TF-123 (all its testcases compile and run successfully), many simple operations involving opaquely differentiable functions are still broken due to TF-1122. I believe that TF-1122 is a distinct problem, so it makes sense to fix it later in a separate PR.

This also adds tests for reabstraction:
- The last 3 tests in `reabstraction_e2e.swift` are the ones that have been fixed by this PR.
- The last 3 tests include some commented-out code that currently segfaults at runtime due to TF-1122.
- All the tests in `reabstraction_e2e.swift` except for the last 3 already pass before this PR. I just used this PR as an opportunity to test reabstraction more thoroughly.

The comment that I added in `AbstractionPattern.h` describes the solution in more detail:

```
    /// A Swift function whose parameters and results are opaque. This is
    /// like `AP::Type<T>((T) -> T)`, except that the number of parameters is
    /// unspecified.
    ///
    /// This is used to construct the abstraction pattern for the
    /// derivative function of a function with opaque abstraction pattern. See
    /// `OpaqueDerivativeFunction`.
    OpaqueFunction,
    /// A Swift function whose parameters are opaque and whose result is the
    /// tuple abstraction pattern `(AP::Opaque, AP::OpaqueFunction)`.
    ///
    /// Purpose. When we reabstract `differentiable_function` using the
    /// `AP::Opaque` pattern, we use `AP::Opaque` to reabstract the original
    /// function in the bundle and we use `AP::OpaqueDerivativeFunction` to
    /// reabstract the derivative functions in the bundle. This preserves the
    /// `differentiable_function` invariant that the derivative type
    /// (`SILFunctionType::getAutoDiffDerivativeFunctionType()`) of the original
    /// function is equal to the type of the derivative function. For example:
    ///
    ///   differentiable_function
    ///     [parameters 0]
    ///     %0 : $@callee_guaranteed (Float) -> Float
    ///     with_derivative {
    ///       %1 : $@callee_guaranteed (Float) -> (
    ///         Float,
    ///         @owned @callee_guaranteed (Float) -> Float
    ///       ),
    ///       %2 : $@callee_guaranteed (Float) -> (
    ///         Float,
    ///         @owned @callee_guaranteed (Float) -> Float
    ///       )
    ///     }
    ///
    /// The invariant-respecting abstraction of this bundle to `AP::Opaque` is:
    ///
    ///   differentiable_function
    ///     [parameters 0]
    ///     %3 : $@callee_guaranteed (@in_guaranteed Float) -> @out Float
    ///     with_derivative {
    ///       %4 : $@callee_guaranteed (@in_guaranteed Float) -> (
    ///         @out Float,
    ///         @owned @callee_guaranteed (@in_guaranteed Float) -> @out Float
    ///       ),
    ///       %5 : $@callee_guaranteed (@in_guaranteed Float) -> (
    ///         @out Float,
    ///         @owned @callee_guaranteed (@in_guaranteed Float) -> @out Float
    ///       )
    ///     }
    ///
    /// In particular:
    ///
    /// - The reabstraction %0 => %3 uses pattern `AP::Opaque`.
    /// - The reabstraction %1 => %4 uses pattern
    ///   `AP::OpaqueDerivativeFunction`, which maximally abstracts all the
    ///   parameters, and abstracts the result as the tuple
    ///   `(AP::Opaque, AP::OpaqueFunction)`.
    /// - The reabstraction %2 => %5 similarly uses pattern
    ///   `AP::OpaqueDerivativeFunction`.
    OpaqueDerivativeFunction,
```

### CODE REVIEW

File: lib/SIL/AbstractionPattern.cpp [hunk 1]
```diff
@@ -869,3 +910,26 @@ const {
     return AbstractionPattern(getGenericSignature(), memberTy);
   }
 }
+
+AbstractionPattern AbstractionPattern::getAutoDiffDerivativeFunctionType(
+    IndexSubset *indices, unsigned resultIndex,
+    AutoDiffDerivativeFunctionKind kind, LookupConformanceFn lookupConformance,
+    GenericSignature whereClauseGenericSignature, bool makeSelfParamFirst) {
```
Comment (dan-zheng) [inline lib/SIL/AbstractionPattern.cpp]:
```suggestion
    GenericSignature derivativeGenericSignature, bool makeSelfParamFirst) {
```

File: include/swift/SIL/AbstractionPattern.h [hunk 2]
```diff
@@ -1020,6 +1126,19 @@ class AbstractionPattern {
...
+  /// Given that the value being abstracted is a function type, return the
+  /// abstraction pattern for the derivative function.
+  ///
+  /// The arguments are the same as the arguments to
+  /// `AnyFunctionType::getAutoDiffDerivativeFunctionType()`.
+  AbstractionPattern getAutoDiffDerivativeFunctionType(
+      IndexSubset *indices, unsigned resultIndex,
+      AutoDiffDerivativeFunctionKind kind,
+      LookupConformanceFn lookupConformance,
+      GenericSignature whereClauseGenericSignature = GenericSignature(),
```
Comment (dan-zheng) [inline include/swift/SIL/AbstractionPattern.h]: Minor: the `whereClauseGenericSignature` argument of `AnyFunctionType::getAutoDiffDerivativeFunctionType` has been renamed to `derivativeGenericSignature` [on `master`](https://github.com/apple/swift/blob/fa1544974b67bcfd2c1cba1023cfb3a381ce7133/include/swift/AST/Types.h#L3260):
```suggestion
      GenericSignature derivativeGenericSignature = GenericSignature(),
```
The renaming will soon propagate to `tensorflow` branch.

File: test/AutoDiff/downstream/reabstraction_e2e.swift [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  expectEqual(Float(7 * 3 * 3), transformed(3))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3, in: transformed))
+}
+
+ReabstractionE2ETests.test("@differentiable function => opaque generic => concrete") {
+  func id<T>(_ t: T) -> T { t }
+  let inner: @differentiable (Float) -> Float = { 7 * $0 * $0 }
+  let transformed = id(inner)
+
+  // TODO(TF-1122): Actually using `transformed` causes a segfault at runtime.
```
Comment (dan-zheng) [inline test/AutoDiff/downstream/reabstraction_e2e.swift]: How about adding a dedicated `not --crash` negative test for TF-1122? In addition to commenting out the crashers in this test.

File: include/swift/SIL/AbstractionPattern.h [hunk 4]
```diff
@@ -238,7 +298,8 @@ class AbstractionPattern {
   static constexpr const unsigned NumOtherDataBits = 28;
   static constexpr const unsigned MaxOtherData = (1 << NumOtherDataBits) - 1;
 
-  unsigned TheKind : 32 - NumOtherDataBits;
+  // SWIFT_ENABLE_TENSORFLOW
+  unsigned TheKind : 33 - NumOtherDataBits;
```
Comment (dan-zheng) [inline include/swift/SIL/AbstractionPattern.h]: Note: I wonder if this change is significant e.g. for upstreaming. Maybe not.

**Review by dan-zheng — APPROVED** [review]
Great solution to a longstanding issue! Nice documentation.

**Review by rxwei — APPROVED** [review]

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

File: lib/SIL/AbstractionPattern.cpp (hunk 1 above)
Comment (marcrasi) [reply lib/SIL/AbstractionPattern.cpp]: done

File: include/swift/SIL/AbstractionPattern.h (hunk 2 above)
Comment (marcrasi) [reply include/swift/SIL/AbstractionPattern.h]: done

File: test/AutoDiff/downstream/reabstraction_e2e.swift (hunk 3 above)
Comment (marcrasi) [reply test/AutoDiff/downstream/reabstraction_e2e.swift]: done

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by dan-zheng** [comment]
`@swift-ci Please test tensorflow` is still down, so I'll help test this PR on macOS manually.

**Comment by dan-zheng** [comment]
I merged `apple/tensorflow` and pushed some minor edits.
`swift/utils/build-script --preset tensorflow_test` passes on macOS. Merging now!

**Pull request merged**